version-upgrade-operator/README.md

```diff
@@ -0,0 +1,491 @@
+deploying an object and our new operator setting the new version), you might need to first commit the target
+object into repository with a flag like `suspend: True` and flip this flag only after the initial version is
+set.
+
+Please also note, that if you're a GitOps user, resolving directly to Flux's
+[auto upgrade capabilities](https://fluxcd.io/flux/guides/image-update/) might be a better solution for you.
+
+### Spec draft and example
+
+For a `trivy` App, we want to update all instances of `App` CRs that have a specific label with the latest
+`rc` release.
+
+The relevant Flux objects used for version discovery (once again, might be changed by an equivalent solution):
+
+#### Version discovery
+
+```yaml
+# potentially ImageRepository CR can be auto-generated
+---
+
+apiVersion: image.toolkit.fluxcd.io/v1beta2
+kind: ImageRepository
+metadata:
+  name: trivy
+spec: 
+  image: myregistry/trivy 
+  interval: 1h 
+  provider: generic 
+status: {}
+
+# ...
+
+lastScanResult: latestTags: - v1.4.2 - v1.4.1 - v1.4.0 - v1.3.1 - multiarch - master-t0dgwgrs -
+master-a9a1252 - latest - ff1fb39 - fed964e scanTime: "2025-04-28T13:01:59Z" tagCount: 227
+
+---
+
+apiVersion: image.toolkit.fluxcd.io/v1beta2 kind: ImagePolicy metadata: name: trivy-rc spec:
+imageRepositoryRef: name: trivy policy: semver: range: ">= 1.0.0" filterTags: pattern: "._-rc._" status:
+
+# ...
+
+status: conditions:
+
+- lastTransitionTime: "2025-04-24T12:16:37Z" message: Latest image tag for 'registry/repo' updated from 0.4.11
+  to 1.1.1 observedGeneration: 2 reason: Succeeded status: "True" type: Ready latestImage: registry/repo:1.1.1
+  observedGeneration: 2 observedPreviousImage: registry/repo:0.4.11
+
+```
+
+#### Upgrade schedule definition
+
+The new CRDs:
+
+```yaml
+# This object is container data only, so doesn't have any status fields
+kind: VersionUpgradeSchedule
+metadata:
+  name: upgrade-stable-at-night
+  namespace: org-testorg
+spec:
+  # validFrom
+  #
+  # Decides the time window when the reconciliation is in power.
+  #
+  # The CR gets reconciled at `.spec.interval` intervals, but upgrading is skipped in
+  # the [-∞, .spec.policy.validFrom) time window. Works like suspending until a given time.
+  #
+  # To suspend indefinitely use
+  # validFrom: "Inf"
+  # TODO: does it run discovery even before "ValidFrom"? I think it should, so we can report that in status
+  # end emit an event.
+  validFrom: "2025-05-05T05:05Z" # time window when reconciliation is in power: [2025.05.05 05:05, ∞)
+
+  ## SCHEDULING options - exclusive, choose just one
+
+  # upgradeWindows
+  #
+  # Defines all the possible upgrade windows. The windows are ORed together, so upgrade will be triggered for any
+  # reconciliation that falls within any of the windows. Within a single window, the conditions are ANDed together,
+  # so `dayOfWeek: Wed ` and timeStart: 7:00` means "only on Wednesdays at 7:00".
+  upgradeWindows:
```

Comment:
as the customer interaction we had a few weeks back mentioned, how would we model sth like "first tuesday of the first full week in a month?" 

related: would cron notation not make it easier as there's existing tooling to convert certain desired patterns into cron which might be harder to model into sth custom like we propose below.